kaos/algorithm/interpolator.py

```diff
@@ -0,0 +1,114 @@
+
+import numpy as np
+from sqlalchemy.orm.exc import NoResultFound
+
+from kaos.models import OrbitRecords, OrbitSegments
+
+class Interpolator:
+    """Utility class to interpolate satellite position and velocity at arbitrary times, 
+    based on existing datapoints.
+    """
+    def __init__(self, platform_id):
+        self.platform_id = platform_id
+
+    @staticmethod
+    def get_segment(platform_id, timestamp):
+        """Get the orbit segment for the platform_id and time.
+
+        Args:
+            platform_id: The UID of the satellite.
+            timestamp: The time to search for.
+
+        Return:
+            The orbit segment for this platform_id that contains the timestamp. If multiple
+            segments include this timestamp, which may occur if timestamp is a segment
+            boundary, then return the segment whose start time is timestamp. If no matching
+            segment exists, return None.
+        """
+        # TODO move this to model
+        try:
+            # pylint: disable=undefined-variable
+            return (OrbitSegments.query.filter(OrbitSegments.platform_id == platform_id,
+                                               OrbitSegments.start_time <= timestamp,
+                                               OrbitSegments.end_time >= timestamp)
+                                        .order_by(OrbitSegments.start_time.desc())
+                                        .first())
+
+            # pylint: enable=undefined-variable
+        except NoResultFound:
+            return None
+
+    @staticmethod
+    def get_orbit_records_by_segment(segment_id):
+        """Get all the orbit records in a segment.
+
+        Args:
+            segment_id: The segment id.
+
+        Return:
+            All the orbit records, sorted by time.
+        """
+        # TODO move this to model
+        try:
+            # pylint: disable=undefined-variable
+            orbit_records = (OrbitRecords.query.filter_by(segment_id=segment_id)
+                                               .order_by(OrbitRecords.time))
+            # pylint: enable=undefined-variable
+            return list(orbit_records)
+        except NoResultFound:
+            return []
+        
+    @staticmethod
+    def linear_interp(platform_id, timestamp):
+        """Use linear interpolation to estimate the position and velocity of a satellite 
+        at a given time.
+
+        Args:
+            platform_id: The UID of the satellite.
+            timestamp: The time for which to get the estimated position and velocity.
+
+        Return:
+            A tuple (pos, vel). Both pos, vel are 3-tuples containing the vector components
+            of the position and velocity. Return None if interpolation could not be done
+            for the given satellite and timestamp.
+        """
+        # find the correct segment
+        segment = Interpolator.get_segment(platform_id, timestamp)
+        if not segment:
+            return None
```

Comment:
IMHO ( we should probably decide on this ), if an error occurs pls throw. This is an irrecoverable situation, meaning if you don't throw the caller ( me ) will throw anyway since there is nothing we can do. Throwing also helps determine the "exact" reason behind the failure.

Reply:
That's fine with me, I was debating the same.

Changed everything to throw ValueErrors. (it turns out NoResultFound is not actually raised by filter_by())